### PR TITLE
feat: custom metadata on upload

### PR DESCRIPTION
**Pull request by fenos**
## What kind of change does this PR introduce?

Feature

## What is the new behavior?

We can now pass custom metadata (up to 1MB) to:

- [x] Standard uploads
- [x] Tus Uploads
- [x] S3 Uploads 

FInally closes https://github.com/supabase/storage/issues/439

### CODE REVIEW

File: migrations/tenant/0025-custom-metadata.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE storage.objects ADD COLUMN user_metadata jsonb NULL;
```
Comment (inian) [inline migrations/tenant/0025-custom-metadata.sql:1]: auth seems to have it as `raw_user_meta_data` but personally think user_metadata reads better

File: migrations/tenant/0025-custom-metadata.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+ALTER TABLE storage.s3_multipart_uploads ADD COLUMN metadata jsonb NULL;
```
Comment (inian) [inline migrations/tenant/0025-custom-metadata.sql]: this one needs to be metadata cos of the s3 protocol? cant we have this also as user_metadata?

**Review by inian — APPROVED** [review]

File: migrations/tenant/0025-custom-metadata.sql (hunk 2 above)
Comment (fenos) [reply migrations/tenant/0025-custom-metadata.sql]: It doesn't have to be `metadata` i can rename it for consistency 👍

Comment (fenos) [reply migrations/tenant/0025-custom-metadata.sql]: Done

**Comment by Gbillington1** [comment]
Thank you!

**Comment by jppty** [comment]
Is there any documentation or example for passing metadata through TUS (e.g. https://github.com/supabase/supabase/blob/master/examples/storage/resumable-upload-uppy/index.html)?

**Pull request merged**